SuiteCarla

emailweixu opened the pull request:
To use this, there two ways:

1. Run the code within docker image horizonrobotics/alf:0.0.3-carla.  Both [Docker](https://docs.docker.com/engine/install/ubuntu/) and [Nvidia-Docker](https://github.com/NVIDIA/nvidia-docker>) need to be installed.

2. Install carla:

```bash
    wget https://carla-releases.s3.eu-west-3.amazonaws.com/Linux/CARLA_0.9.9.tar.gz
    mkdir carla
    tar zxf CARLA_0.9.9.tar.gz -C carla
    cd carla/Import
    wget https://carla-releases.s3.eu-west-3.amazonaws.com/Linux/AdditionalMaps_0.9.9.tar.gz
    cd ..
    ./ImportAssert.sh
    easy_install PythonAPI/carla/dist/carla-0.9.9-py3.7-linux-x86_64.egg
```

The unittest can be run in interactive mode using:
```bash
python suite_carla_test.py --manual 1
```
Make sure you are using python3.7

Relevant Carla documentation: 
1. [Python API](https://carla.readthedocs.io/en/latest/python_api/)
2. [Carla Sensors](https://carla.readthedocs.io/en/latest/ref_sensors/)

emailweixu committed:
SuiteCarla

hnyu commented:
> 1. Run the code within docker image horizonrobotics/alf:0.0.3-carla.  Both [Docker](https://docs.docker.com/engine/install/ubuntu/) and [Nvidia-Docker2](https://github.com/NVIDIA/nvidia-docker%3E) need to be installed.

Why is the version of 0.0.3? Can we actually just overwrite the current alf:0.0.5 by including carla it?

emailweixu commented:
> > 1. Run the code within docker image horizonrobotics/alf:0.0.3-carla.  Both [Docker](https://docs.docker.com/engine/install/ubuntu/) and [Nvidia-Docker2](https://github.com/NVIDIA/nvidia-docker%3E) need to be installed.
> 
> Why is the version of 0.0.3? Can we actually just overwrite the current alf:0.0.5 by including carla it?

On hub.docker.com the latest version is alf:0.0.3-pytorch. I didn't change alf:0.0.3-pytorch to include carla for two reasons, 1, alf:0.0.3-pytorch is a CPU image;2, alf:0.0.5-carla is much bigger (8.81GB) vs. 1.4GB.

We need to build a different image for internal cluster use.

emailweixu committed:
Fix _get_unused_port bug

emailweixu committed:
Update comments

hnyu commented:
When I launch the docker image, the terminal hangs with the following messages:

`4.24.3-0+++UE4+Release-4.24 518 0
Disabling core dumps.
sh: 1: xdg-user-dir: not found
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default`

Any idea of what's going on?

emailweixu commented:
> When I launch the docker image, the terminal hangs with the following messages:
> 
> `4.24.3-0+++UE4+Release-4.24 518 0 Disabling core dumps. sh: 1: xdg-user-dir: not found ALSA lib confmisc.c:767:(parse_card) cannot find card '0' ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default ALSA lib confmisc.c:767:(parse_card) cannot find card '0' ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default`
> 
> Any idea of what's going on?

This is ok. I have these errors too. Doesn't seem to affect running.

hnyu commented:
> This is ok. I have these errors too. Doesn't seem to affect running.

But the terminal hangs. I can't type anything.

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+# limitations under the License.
+"""CarlaEnvironment suite.
+
+To use this, there two ways:
```
there are two ways

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+        self._parent = parent_actor
+
+    def destroy(self):
+        """The the commands for destroying this sensor..
```
Return the commands for ...

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+    def destroy(self):
+        """The the commands for destroying this sensor..
+
+        Use carla.Client.apply_batch_sync() to actually destroy the sensor.
```
\``carla.Client.apply_batch_sync()\``

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+            current_frame (int): current frame no. For some sensors, they may
+                not receive any data in the most recent tick. ``current_frame``
+                will be compared against the frame no of the last received data
+                to make sure no data is correctly interpretted.
```
"no data" -> "data"

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+    def observation_desc(self):
+        return (
+            "Impulses from collision during the last tick. Each impulse is "
+            "a 3-D vector. At most %d collisions are used. The result is padded "
```
Maybe explain what the 3D vector means.

emailweixu replied:
Carla documentation doesn't say what exactly is. I guess it's the force vector from the impulse.

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+class CollisionSensor(SensorBase):
+    """CollisionSensor for getting collision signal.
+
+    It get the impulses from the collisions during the last tick.
```
Can it also return what object the car collides with?

emailweixu replied:
In `` _on_collision()`` ``event.other_actor`` is the carla.Actor object that it collides with. It's not obvious how to encode this though. We can leave it as a TODO.

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+            "with zeros if there are less than %d collisions" %
+            (self._max_num_collisions, self._max_num_collisions))
+
+    def get_current_observation(self, current_frame):
```
comments to show in docs

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+    """LaneInvasionSensor for detecting lane invasion.
+
+    Lane invasion cannot be directly observed by raw sensors used by real cars.
+    So main purpose of this is to provide training signla (e.g. reward).
```
signal

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+    def observation_desc(self):
+        return "A vector of [latitude (degrees), longitude (degrees), altitude (meters to be confirmed)]"
+
+    def get_current_observation(self, current_frame):
```
Maybe copy the return of `observation_desc` as a docstring here

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+            "velocity in rad/s^2, and compass is orientation with regard to the "
+            "North ((0.0, 1.0, 0.0) in Unreal Engine) in radians.")
+
+    def get_current_observation(self, current_frame):
```
docstring

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+        self._sensor = world.spawn_actor(
+            bp,
+            carla.Transform(
+                carla.Location(x=2.8, z=1.0), carla.Rotation(pitch=5)),
```
Do we need to make all the parameters changeable in this init function?

emailweixu replied:
Good idea.

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+            "angle of the detection in radians, and depth id the distance from "
+            "the sensor to the detection in meters.")
+
+    @property
```
The other `get_current_observation` is a function instead of a property; docstring.

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+        self._image = np.zeros([num_channels, image_size_y, image_size_x],
+                               dtype=np.uint8)
+
+    def render(self, display):
```
comments

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+    def observation_spec(self):
+        return self._observation_spec
+
+    def get_current_observation(self, current_frame):
```
docstring

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+
+
+class Player(object):
+    """Player.
```
Remove "Player."

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+        return self._observation_desc
+
+    def action_spec(self):
+        """Get the action spec.
```
Docstring explaining the actions

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+        Args:
+            current_frame (int): current simulation frame no.
+        Returns:
+            TimeStep: all elements are nd.ndarray or np.number.
```
\``np.ndarray\`` or \``np.number\``

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+                self._current_distance / self._min_speed / self._delta_seconds)
+            self._fail_frame = current_frame + max_frames
+        elif (self._current_distance < self._success_distance_thresh
+              and self._actor.get_velocity() == carla.Location()):
```
should get_velocity() be get_location() instead?

emailweixu replied:
get_velocity() is correct carla.Location() is same as carla.Location(0.,0.,0.). The purpose is to check if the vehicle is not moving

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+              and self._actor.get_velocity() == carla.Location()):
+            # TODO: include waypoint orientation as success critiria
+            self._done = True
+            step_type = ds.StepType.LAST
```
In this case the discount should be 0? Otherwise the step_type should still be MID. For the former case, need to set self._fail_frame = current_frame.

emailweixu replied:
Added discount=0. I don't understand why need to set self._fail_frame=current_frame

hnyu replied:
You're right. fail_frame will be set to None when resetting

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+    def act(self, action):
+        """Generate the carla command for taking the given action.
+
+        Use carla.Client.apply_batch_sync() to actually destroy the sensor.
```
\``carla.Client.apply_batch_sync()\``

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+        reward = 0.
+        if self._fail_frame is None:
+            step_type = ds.StepType.FIRST
+            max_frames = math.ceil(
```
So basically this means that we cannot manually set a max_episode_steps in a gin file?

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+        Args:
+            mode (str): one of ['rgb', 'human']
+        Returns
+            None: if mode is 'human'
```
tuple:
\- None: ...
\- np.ndarray: ...

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+                 docker_image="horizonrobotics/alf:0.0.3-carla",
+                 quality_level="Low",
+                 carla_root="/home/carla"):
+        """Start a Carla server.
```
Remove this line

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+                value is correct for using docker image. If not using docker
+                image, make sure you provide the correct path. This is the directory
+                where you unzipped the file you downloaded from
+                https://github.com/carla-simulator/carla/releases/tag/0.9.9.
```
\`<>\`_

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+            blueprints
+        ) > 0, "Cannot find safe vehicle '%s'" % self._vehicle_filter
+
+        spawn_points = self._world.get_map().get_spawn_points()
```
Each map only has a fixed number of spawn points?

emailweixu replied:
correct.

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+                color = random.choice(
+                    blueprint.get_attribute('color').recommended_values)
+                blueprint.set_attribute('color', color)
+            if blueprint.has_attribute('driver_id'):
```
What's the purpose of having random driver ids?

emailweixu replied:
Actually, I don't know what effect driver id has. Cannot find documentation about it. Carla source code doesn't say it either. I will add a TODO to figure it out.

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1163 @@
+        self._server.stop()
+
+    @property
+    def players(self):
```
docstring

hnyu commented on alf/environments/suite_carla_test.py:
```diff
@@ -0,0 +1,162 @@
+import torch
+
+import alf
+from alf.environments import gym_wrappers, suite_carla, torch_environment
```
gym_wrappers is not needed

emailweixu committed:
Address comments

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -79,9 +79,9 @@ def __init__(self, parent_actor):
 
     def destroy(self):
-        """The the commands for destroying this sensor..
+        """Return the the commands for destroying this sensor.
```
remove 'the'

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -342,10 +366,15 @@ def get_current_observation(self, current_frame):
+                 max_num_detections=200):
         """
         Args:
             parent_actor (carla.Actor): the parent actor of this sensor.
```
add arg comments about 'xyz' and 'pyr'

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -0,0 +1,1233 @@
+        else:
+            assert os.path.exists(carla_root + "/CarlaUE4.sh"), (
+                "%s/CarlaUE4.sh "
+                "does not exist. Please provide correct value for `carla_root`"
```
% carla_root

hnyu commented:
When the test case ends, there was a warning message:

> /usr/lib/python3.7/subprocess.py:883: ResourceWarning: subprocess 13306 is still running
>  ResourceWarning, source=self)
> ResourceWarning: Enable tracemalloc to get the object allocation traceback

Is this OK?

emailweixu commented:
> When the test case ends, there was a warning message:
> 
> > /usr/lib/python3.7/subprocess.py:883: ResourceWarning: subprocess 13306 is still running
> > ResourceWarning, source=self)
> > ResourceWarning: Enable tracemalloc to get the object allocation traceback
> 
> Is this OK?

Fixed. Need to call process.communicate() after kill

emailweixu committed:
Address comments

Add __del__() for CarlaServer and CarlaEnvironment to make resource release more robust.
